### PR TITLE
Support Expert Parallelism

### PR DESCRIPTION
**Pull request by NouamaneTazi**

### CODE REVIEW

File: src/nanotron/config/parallelism_config.py [hunk 1]
```diff
@@ -20,6 +20,7 @@ class ParallelismArgs:
         dp: Number of DP replicas
         pp: Number of PP stages
         tp: Number of TP replicas
+        expert_parallel_size: Number of expert parallel replicas (used only for MoEs)
```
Comment (xrsrke) [inline src/nanotron/config/parallelism_config.py:23]: Isn't expert_parallel_size should be the number of experts per tp rank?

File: src/nanotron/config/parallelism_config.py [hunk 2]
```diff
@@ -34,6 +35,8 @@ class ParallelismArgs:
     recompute_granularity: Optional[RecomputeGranularity] = None
     tp_linear_async_communication: Optional[bool] = None
 
+    expert_parallel_size: Optional[int] = None
```
Comment (xrsrke) [inline src/nanotron/config/parallelism_config.py:38]: 1 by default?

File: src/nanotron/parallel/context.py [hunk 3]
```diff
@@ -16,9 +16,10 @@ def __init__(
         pipeline_parallel_size: int,
         data_parallel_size: int,
         backend: DistributedBackend = "nccl",
+        expert_parallel_size: int = 1,
```
Comment (xrsrke) [inline src/nanotron/parallel/context.py:19]: Move it after `data_parallel_size` (parallel_size..., backend)

File: src/nanotron/trainer.py [hunk 4]
```diff
@@ -198,7 +199,7 @@ def __init__(
 
         # Setup tensorboard write and log writers on output rank
         self.logger_ranks = self.parallel_context.world_rank_matrix[
-            self.unwrapped_model.output_pp_rank, 0, 0
+            0, self.unwrapped_model.output_pp_rank, 0, 0
```
Comment (xrsrke) [inline src/nanotron/trainer.py:202]: would be nice if we do self.parallel_context.world_rank_matrix(expert_rank=0, pp_rank=0, tp_rank=0, dp_rank=0)

File: src/nanotron/trainer.py [hunk 5]
```diff
@@ -350,17 +351,8 @@ def training_step(
...
-                mp_pg=self.parallel_context.world_ranks_to_pg[
-                    tuple(
-                        sorted(
-                            self.parallel_context.world_rank_matrix[
-                                :, dist.get_rank(self.parallel_context.dp_pg), :
-                            ].reshape(-1)
-                        )
-                    )
-                ],
+                mp_pg=self.parallel_context.mp_pg,
```
Comment (xrsrke) [inline src/nanotron/trainer.py:355]: nice!

**Review by xrsrke — CHANGES_REQUESTED** [review]

File: src/nanotron/config/parallelism_config.py (hunk 1 above)
Comment (NouamaneTazi) [reply src/nanotron/config/parallelism_config.py:23]: Not quite, expert parallelism is orthogonal to TP. for example you can have 1 expert sharded along 2 tp ranks

File: src/nanotron/config/parallelism_config.py (hunk 2 above)
Comment (NouamaneTazi) [reply src/nanotron/config/parallelism_config.py:38]: true

File: src/nanotron/trainer.py (hunk 4 above)
Comment (NouamaneTazi) [reply src/nanotron/trainer.py:202]: i very much like this! but that could be in an othogonal PR

Comment (NouamaneTazi) [reply src/nanotron/trainer.py:202]: added a TODO

**Pull request merged**